sdk/python/feast/infra/registry/caching_registry.py

```diff
@@ -289,22 +290,32 @@ def refresh(self, project: Optional[str] = None):
         self.cached_registry_proto_created = datetime.utcnow()
 
     def _refresh_cached_registry_if_necessary(self):
-        with self._refresh_lock:
-            expired = (
-                self.cached_registry_proto is None
-                or self.cached_registry_proto_created is None
-            ) or (
-                self.cached_registry_proto_ttl.total_seconds()
-                > 0  # 0 ttl means infinity
-                and (
-                    datetime.utcnow()
-                    > (
-                        self.cached_registry_proto_created
-                        + self.cached_registry_proto_ttl
+        if self.cache_mode == "sync":
+            with self._refresh_lock:
+                expired = (
+                    self.cached_registry_proto is None
+                    or self.cached_registry_proto_created is None
+                ) or (
+                    self.cached_registry_proto_ttl.total_seconds()
+                    > 0  # 0 ttl means infinity
+                    and (
+                        datetime.utcnow()
+                        > (
+                            self.cached_registry_proto_created
+                            + self.cached_registry_proto_ttl
+                        )
                     )
                 )
-            )
 
-            if expired:
-                logger.info("Registry cache expired, so refreshing")
-                self.refresh()
+                if expired:
+                    logger.info("Registry cache expired, so refreshing")
+                    self.refresh()
+
+    def _start_thread_async_refresh(self, cache_ttl_seconds):
+        self.registry_refresh_thread = threading.Timer(cache_ttl_seconds, self.refresh)
```

Comment:
This will just run once, won't it? I think you meant to call refresh first and spawn a thread calling `_start_thread_async_refresh` to create an infinite loop. Also you shouldn't start a thread if `cache_ttl_seconds == 0`

Reply:
> This will just run once, won't it? I think you meant to call refresh first and spawn a thread calling `_start_thread_async_refresh` to create an infinite loop. Also you shouldn't start a thread if `cache_ttl_seconds == 0`

@tokoko Threading.Timer will spin up a new thread and run a function at specified intervals hence it does not only run once. It is an infinite loop.  `cache_ttl_seconds > 0` implemented

Reply:
Can you doublecheck? Just tried running this `threading.Timer(2, lambda: print('Hello World')).start()` and it only ran once. docs don't mention anything about Timer creating an infinite loop either. Take a look at how it's implemented in [feature server](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L77).

Reply:
@tokoko you are correct, it is executed once. I made adjustments to run in an infinite loop.